File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
# 최대한 빨리 도착 -> 최단거리 -> bfs
# nxm 크기 맵
# maps: 0은 벽이 있는 자리, 1은 벽이 없는 자리
from collections import deque

def solution(maps):
    answer = 0
    # 상하좌우 이동
    dx = [0, 0, -1, 1]
    dy = [1, -1, 0, 0]
    
    # bfs
    def bfs(x, y):
        # 1. 시작노드를 큐에 삽입, 방문처리
        q = deque()
        q.append((x,y))
        
        while q: # 큐가 빌때까지 반복
            x, y = q.popleft()
            # 상하좌우 칸 확인
            
            for i in range(4):
                nx = x+dx[i]
                ny = y+dy[i]
                
                if 0<= nx< len(maps) and 0<=ny< len(maps[0]) and maps[nx][ny]==1:
                    maps[nx][ny] = maps[x][y] +1 # 거리 증가
                    q.append((nx,ny)) # 큐에 삽입
                    
        return maps[len(maps)-1][len(maps[0])-1]
                    
    answer = bfs(0,0)
           
    # answer가 1 그대로 -> 도달 못함!
    if answer==1:
        return -1
    else:
        return answer

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
// 5x5 크기 맵, 동/서/남/북 이동, 게임 맵 벗어난 길은 불가
// 상대팀 진영에 도착할 수 없을 땐 -1 반환
// 0: 벽, 1: 벽x
// 최단거리-bfs
import java.util.*;

class Solution {
    static int[]  dx = {-1, 0, 1, 0};
    static int[]  dy = {0, 1, 0, -1};
    static boolean[][] visited;
    
    public int solution(int[][] maps) {
        int answer = 0;
        int n = maps.length; 
        int m = maps[0].length;
        
        // bfs-queue 사용
        Queue<int[]> qu = new LinkedList<>();
        visited= new boolean[n][m]; // 해당 위치 방문 체크
        
        qu.offer(new int[]{0,0,1}); // x,y,현재까지 거리
        visited[0][0] = true;
        
        while(!qu.isEmpty()){
            int[] cur = qu.poll();
            int x = cur[0];
            int y = cur[1];
            int d = cur[2];
            // 도착했다면 거리 반환
            if(x==n-1 && y==m-1){
                return d;
            }
            for(int i=0; i<4; i++){
                int nx = x+dx[i];
                int ny = y+dy[i];
                // 범위 체크, 방문 체크, 벽 x
                if(nx>=0 && nx<n && ny>=0 && ny<m && !visited[nx][ny] && maps[nx][ny]==1){
                    qu.offer(new int[]{nx, ny, d+1});
                    visited[nx][ny]=true;
                }
            }
        }
        
        return -1;
    }
}